README.md

```diff
@@ -56,6 +56,31 @@ const results = await conn.execute('select 1 from dual')
 console.log(results)
 ```
 
+### Transactions
+
+Use the `transaction` function to safely perform database transactions. If any unhandled errors are thrown during execution of the transaction, the transaction will be rolled back.
+
+The following example is based on [the Slotted Counter Pattern](https://planetscale.com/blog/the-slotted-counter-pattern).
+
+```ts
+import { connect } from '@planetscale/database'
+
+const config = {
+  host: '<host>',
+  username: '<user>',
+  password: '<password>'
+}
+
+const conn = connect(config)
+const results = await conn.transaction((tx) => {
+  return Promise.all([
+    tx.execute('INSERT INTO branches (database_id, name) VALUES (?, ?), [42, "planetscale"]'),
+    tx.execute('INSERT INTO slotted_counters(record_type, record_id, slot, count) VALUES (branch_count, database_id, RAND() * 100, 1) ON DUPLICATE KEY UPDATE count = count + 1')
+  ])
```

Comment:
I'm not smart with this stuff. Is this safe to consider parallel queries within a transaction?

I'd likely not really encourage this since these might need to be sequential?

But I'm really not sure what's expected here.

Reply:
I'm still not sure about promoting parallel execution since the behavior is nondeterministic. This is very atypical expectations of a driver, and I know our APIs enable it. I'm just not sure it's a good idea to encourage. This is a lot more clear when you have 1:1 mapping between connections and sessions. You cannot execute multiple concurrent queries over a single mysql connection, so the problem doesn't actually exist. If you did parallel queries, you'd have multiple sessions/connections. So this just gets into a pretty grey area since the sessions might diverge based on the outcome of the parallel queries, and since we overwrite the internal state on each response, this ends up being like, a last query wins the state, which is nondeterministic.

Reply:
Yeah, the more I'm thinking about this, the more I dislike it. A more safe, analagous way, would be something like:

```javascript
const db = new Client({...})

Promise.all([
  db.connection().execute(...),
  db.connection().execute(...),
])
```

I'm not sure how this works with transactions though since I think there's an assumption of linearalizability within a transaction and parallel queryies within a single one seems... not good.

I'm going to leave this to yall, but I think we shouldnt' encourage this behavior at least. If folks want to contrive a way to do it, sure, but we shouldn't encourage shoving everything into one promise to execute a bunch of queries. We should execute them serially.

but this is wehre I think there'd be value in us adding an `/ExecuteMany` API endpoint, so multiple can be sent, and receive multiple results in serial with 1 round trip. That way they're executed as would be predictable, without the overhead of going back and forth.

Reply:
What does the server do when it receives two `post` requests for the same session token simultaneously? Does it queue them and execute them serially on the mysql connection?

Reply:
I'm not actually sure, this would be a vitess question. We just are using the vtgate Execute RPC endpoint, so this is all pretty much offloaded from us.